### PR TITLE
Dynamic concept properties

### PR DESCRIPTION
**Pull request by MiroslavGatsanoga**
# Description

## What

Define dynamic concept properties in the ontology package.

## Why

[JIRA ticket](https://financialtimes.atlassian.net/browse/UPPSF-2468).

## Anything, in particular, you'd like to highlight to reviewers

Mention here sections of code which you would like reviewers to pay extra attention to .E.g

This is branched from https://github.com/Financial-Times/concepts-rw-neo4j/pull/72, review the commits from `Separate creation of node queries for source and canonical` onward.

## Scope and particulars of this PR (Please tick all that apply)

- [ ] Tech hygiene (dependency updating & other tech debt)
- [ ] Bug fix
- [X] Feature
- [ ] Documentation
- [ ] Breaking change
- [ ] Minor change (e.g. fixing a typo, adding config)

___
This Pull Request follows the rules described in our [Pull Requests Guide](https://github.com/Financial-Times/upp-docs/tree/master/guides/pr-guide)

### CODE REVIEW

File: ontology/concept.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
+package ontology
+
+func (c NewAggregatedConcept) GetPropertyValue(propName string) (interface{}, bool) {
+	val, found := c.Properties[propName]
+	if !found {
+		return nil, false
+	}
+
+	switch v := val.(type) {
```
Comment (ivan-p-nikolov) [inline ontology/concept.go:9]: **thought:** I don't know how I feel about the idea that the `getter` is also a `validator`. It's fine solution, but something is off, and I don't know what.

File: ontology/concept.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+	case []interface{}:
```
Comment (ivan-p-nikolov) [inline ontology/concept.go:10]: We talked about it, but just to have it in writing. When we start transitioning properties we would need a way to validate that the stored property is of the type defined in the onthology config.

File: ontology/transform.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+import "encoding/json"
+
+func TransformToNewAggregateConcept(old AggregatedConcept) NewAggregatedConcept {
+	var newSources []NewConcept
+	for _, s := range old.SourceRepresentations {
+		newSources = append(newSources, TransformToNewSourceConcept(s))
+	}
+
+	oldMap := map[string]interface{}{}
+	oldBytes, _ := json.Marshal(old)
```
Comment (ivan-p-nikolov) [inline ontology/transform.go:12]: I love the simplicity of the `Marshal/Unmarshal` solution to map from old to new concept.

File: ontology/transform.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+	_ = json.Unmarshal(oldBytes, &oldMap)
+
+	props := map[string]interface{}{}
+	for field := range FieldToNeoProps {
```
Comment (ivan-p-nikolov) [inline ontology/transform.go]: Am I understanding correctly that the property names in memory will be the ones in Neo4j and not the "public" ones? Meaning in the properties map we will have `imageUrl` and not `_imageUrl`?

**Review by ivan-p-nikolov — APPROVED** [review]

File: ontology/concept.go (hunk 1 above)
Comment (MiroslavGatsanoga) [reply ontology/concept.go:9]: Yeah, I know that the method does two things but currently we always want to validate when getting property value so I decided to make it a single function. If the need to validate only without getting the value arises in the future we can split the function.

File: ontology/transform.go (hunk 4 above)
Comment (MiroslavGatsanoga) [reply ontology/transform.go]: Not sure if I understand the question, here is an example `config.yml` to ilustrate what we'll have in `FieldToNeoProps`:
```
fieldToNeoProps:
  aliases: "aliases"
  emailAddress: "emailAddress"
  _imageUrl: "imageUrl
```

File: Dockerfile.tests [hunk 5]
```diff
@@ -5,12 +5,17 @@ ENV PROJECT=concepts-rw-neo4j
...
 WORKDIR ${SRC_FOLDER}
 
 RUN echo "Fetching dependencies..." \
   && git clone https://github.com/vishnubob/wait-for-it.git \
   && cd wait-for-it \
   && mv ./wait-for-it.sh ${SRC_FOLDER}
 
+COPY go.mod /
+COPY go.sum /
+RUN go mod download
```
Comment (ivanruski) [inline Dockerfile.tests:17]: **question:** What is the reason to copy `go.mod` and `go.sum` to the base path `/` when your working directory is `/go/src/github.com/Financial-Times/concepts-rw-neo4j`

**Review by ivanruski — APPROVED** [review]

File: Dockerfile.tests (hunk 5 above)
Comment (MiroslavGatsanoga) [reply Dockerfile.tests:17]: No particular reason, it doesn't matter from which directory `go mod download` is run as its purpose is to populate the local modules cache. Once populated we leverage that with Docker image layer caching to speed up the testing process locally.

**Pull request merged**